### PR TITLE
Jetpack CP: display Jetpack benefits bottom banner for JCP sites and 5 days after the last dismissal

### LINKED ISSUE #5368: Jetpack CP: display Jetpack benefits bottom banner
Dependencies:
- UI https://github.com/woocommerce/woocommerce-ios/issues/5362
- Backend https://github.com/woocommerce/woocommerce-ios/issues/5364

- [ ] Persist the state when the user taps to dismiss the banner so that it doesn't show up anymore for the site (they can still access it in Settings) (we can use `GeneralStoreSettings` via `AppSettingsAction`)
- [ ] Display the Jetpack benefits bottom banner in the My Store tab when:
  - the site's `isJetpackCPConnected` is `true`
  - the banner was last dismissed at least 5 days ago (matching Android behavior in https://github.com/woocommerce/woocommerce-android/pull/4995)
  - JCP feature flag is enabled

### PR DESCRIPTION
**Pull request by jaclync**
Closes: #5368 

### Description

This PR implements the logic to show/hide the Jetpack benefits banner in the My store tab (it's always shown behind the JCP flag before). The main changes include two parts:

- Persist the state when the user taps to dismiss the banner so that it doesn't show up anymore for the app for 5 calendar days (they can still access the Install Jetpack CTA in Settings) following the [Android implementation](https://github.com/woocommerce/woocommerce-android/pull/4995)
  - Changes in `GeneralAppSettings`, `AppSettingsAction`, and `AppSettingsStore`
- Display the Jetpack benefits bottom banner in the My Store tab when:
  - the site's `isJetpackCPConnected` is `true`
  - JCP feature flag is enabled
  -  the banner was last dismissed at least 5 days ago

Then `DashboardViewController`:
- set the last dismissed date when the user taps to dismiss the Jetpack banner
- observe the app's default `Site` and dashboard UI content view, then load the app settings to determine whether to display the Jetpack banner 

### Testing instructions

Prerequisite: the account is connected to at least a JCP site (setup tips in p1636091588189400-slack-C6H8C3G23) and a Jetpack site (with Jetpack-the-plugin)

Note: I recommend testing on a device to test the "5 days later" dismissal behavior. Updating Mac's system time to test on a simulator might mess up with some services, like time tracker for me 😅 If you run into provisioning profile errors (a new entitlement was added recently for associated domains), you can download the latest profiles following PaaHJt-2Ml-p2

#### JCP feature flag enabled

- Launch the app and log in to a Jetpack site if needed --> there should be no Jetpack banner at the bottom of My store tab
- Go to Settings > Switch Store, and switch to a JCP site --> there should be a Jetpack banner at the bottom of My store tab
- Tap the X button to dismiss the banner --> the banner should disappear
- Relaunch the app --> the banner is still hidden for the site
- In system settings, update the date & time to 4 full calendar days later then relaunch the app --> the banner is still hidden
- In system settings, update the date & time to 5 full calendar days later then relaunch the app --> the banner should be shown again!

#### JCP feature flag disabled

- In `DefaultFeatureFlagService`, return `false` for `jetpackConnectionPackageSupport` feature flag
- Launch the app while still being logged in to a JCP site from the previous test case --> there should be no Jetpack banner at the bottom of My store tab

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

File: Yosemite/YosemiteTests/Stores/AppSettingsStoreTests.swift [hunk 1]
```diff
@@ -519,6 +519,127 @@ final class AppSettingsStoreTests: XCTestCase {
         XCTAssertTrue(isEnabled)
     }
 
+    func test_loadJetpackBenefitsBannerVisibility_returns_true_on_new_generalAppSettings() throws {
+        // Given
+        try fileStorage?.deleteFile(at: expectedGeneralAppSettingsFileURL)
+
+        let currentTime = Date(timeIntervalSince1970: 1638105924)
```
Comment (itsmeichigo) [inline Yosemite/YosemiteTests/Stores/AppSettingsStoreTests.swift:528]: Minor nit: It would be great to add a comment explaining that just any random time here should work. I was a bit confused at first about when this time is 😅

File: Yosemite/YosemiteTests/Stores/AppSettingsStoreTests.swift [hunk 2]
```diff
@@ -519,6 +519,127 @@ final class AppSettingsStoreTests: XCTestCase {
...
+            self.subject?.onAction(action)
+        }
+
+        // Then
+        XCTAssertTrue(isVisible)
+    }
+
+    /// Tests an edge case where the time interval since the last dismissed date is less than 5 24-hour days, but is exactly 5 days on calendar with daily
+    /// saving time.
+    func test_loadJetpackBenefitsBannerVisibility_returns_false_after_setting_last_dismissed_date_exactly_five_24hr_days_ago() throws {
```
Comment (itsmeichigo) [inline Yosemite/YosemiteTests/Stores/AppSettingsStoreTests.swift:579]: Thanks for considering this edge case too! Timezones are so hard especially when there's DST involved 😰

File: Yosemite/Yosemite/Stores/AppSettingsStore.swift [hunk 3]
```diff
@@ -749,6 +781,7 @@ enum AppSettingsStoreErrors: Error {
     case noProductsSettings
     case writeProductsSettings
     case noEligibilityErrorInfo
+    case noLastJetpackBenefitsBannerDismissedTime
```
Comment (itsmeichigo) [inline Yosemite/Yosemite/Stores/AppSettingsStore.swift]: I think this case is not used anywhere - should it be removed?

**Review by itsmeichigo — APPROVED** [review]
Great work! I left a few nits and also have a couple of questions:
- The check only works when the app is relaunched, this is expected right? I suppose that's good enough because we don't want to run timer to check - that would be an overkill 😅.
- For this case: 
> JCP feature flag disabled: Launch the app and log in to a JCP site if needed --> there should be no Jetpack banner at the bottom of My store tab

I could not confirm because my JCP site doesn't show up on the site picker with the feature flag off. I think this is expected but just wanna leave a comment to make sure 😄.

File: Yosemite/YosemiteTests/Stores/AppSettingsStoreTests.swift [hunk 4]
```diff
@@ -519,6 +519,127 @@ final class AppSettingsStoreTests: XCTestCase {
...
+        // Given
+        try fileStorage?.deleteFile(at: expectedGeneralAppSettingsFileURL)
+
+        // GMT - Tuesday, November 23, 2021 1:25:24 PM
+        let lastDismissedTime = Date(timeIntervalSince1970: 1637673924)
+        // GMT - Sunday, November 28, 2021 1:25:24 PM - exactly five days after the last dismissed date without DST
+        let currentTime = Date(timeIntervalSince1970: 1638105924)
+        let calendar: Calendar = {
+            var calendar = Calendar(identifier: .gregorian)
+            guard let timeZoneWithoutDailySavingTime = TimeZone(identifier: "Asia/Taipei") else {
```
Comment (selanthiraiyan) [inline Yosemite/YosemiteTests/Stores/AppSettingsStoreTests.swift]: Super nit: Are we referring to Daylight Saving Time as "DailySavingTime" here? I am wondering whether the term "daylight saving time" is valid. 🤔 
Please ignore if I am wrong. 😅

Comment (itsmeichigo) [reply Yosemite/YosemiteTests/Stores/AppSettingsStoreTests.swift]: Good catch! I think it's just a typo 😛

Comment (jaclync) [reply Yosemite/YosemiteTests/Stores/AppSettingsStoreTests.swift]: Great catch on the typo @selanthiraiyan! I'm not sure why I was thinking "daily saving time" 😅

File: Yosemite/YosemiteTests/Stores/AppSettingsStoreTests.swift (hunk 1 above)
Comment (jaclync) [reply Yosemite/YosemiteTests/Stores/AppSettingsStoreTests.swift:528]: Good point! Added some comments in https://github.com/woocommerce/woocommerce-ios/pull/5505/commits/5d82aedc78b7120b7d8d1845750712195832e9d7

File: Yosemite/YosemiteTests/Stores/AppSettingsStoreTests.swift (hunk 4 above)
Comment (jaclync) [reply Yosemite/YosemiteTests/Stores/AppSettingsStoreTests.swift]: Fixed in https://github.com/woocommerce/woocommerce-ios/pull/5505/commits/5d82aedc78b7120b7d8d1845750712195832e9d7

File: Yosemite/Yosemite/Stores/AppSettingsStore.swift (hunk 3 above)
Comment (jaclync) [reply Yosemite/Yosemite/Stores/AppSettingsStore.swift]: Yup this isn't needed anymore! removed in https://github.com/woocommerce/woocommerce-ios/pull/5505/commits/5d82aedc78b7120b7d8d1845750712195832e9d7

**Comment by jaclync** [comment]
Thank you for catching a few things @itsmeichigo @selanthiraiyan! 🙌  See if you'd like to take another look 😄 

Answering @itsmeichigo's questions:

> The check only works when the app is relaunched, this is expected right? I suppose that's good enough because we don't want to run timer to check - that would be an overkill 😅.

yes, the visibility is only updated whenever the dashboard content view is updated (when the dashboard stats eligibility changes) and the site is loaded (on app launch or after switching stores). As you said, it doesn't seem worthwhile to regularly check the time just to show the Jetpack banner 🙂 

> For this case:
JCP feature flag disabled: Launch the app and log in to a JCP site if needed --> there should be no Jetpack banner at the bottom of My store tab
> I could not confirm because my JCP site doesn't show up on the site picker with the feature flag off. I think this is expected but just wanna leave a comment to make sure 😄.

Ah I was testing rebuilding the app while being logged in to a JCP site, so I still had access to the JCP site. I updated the testing steps for future reference!
